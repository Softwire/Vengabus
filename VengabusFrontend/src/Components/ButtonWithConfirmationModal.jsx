import React from 'react';
import { Modal, Alert, Button, Tooltip, OverlayTrigger } from 'react-bootstrap';

const defaultState = {
    show: false
};

class ButtonWithConfirmationModal extends React.Component {
    constructor(props) {
        super(props);

        this.state = defaultState;
    }

    handleOpen = () => {
        const afterOpen = (this.props.afterShowModalAction || (() => { }));
        this.setState({ show: true }, afterOpen);
    }

    handleConfirm = () => {
        this.props.confirmAction();
        this.handleClose();
    }

    handleClose = () => {
        const afterClose = (this.props.afterCloseModalAction || (() => { }));
        this.setState(defaultState, afterClose);
    }

    render() {
        const style = this.props.buttonDisabled ? {
            pointerEvents: 'none'
        } : {};

        const tooltip = (
            <Tooltip id="tooltip">
                <strong>{this.props.tooltipMessage}</strong>
            </Tooltip>
        );

        const button =
            <Button
                onClick={this.handleOpen}
                bsStyle={this.props.buttonStyle || "danger"}
                disabled={!!this.props.buttonDisabled}
                style={style}
            >
                {this.props.buttonText}
            </Button>;

        const buttonWithToolTip =
            <OverlayTrigger rootClose overlay={tooltip} placement="top">
                <div style={{ display: 'inline-block', cursor: 'not-allowed', float: 'right' }}>
                    {button}
                </div>
            </OverlayTrigger>;


        return (
            <React.Fragment>
                {this.props.tooltipMessage ? buttonWithToolTip : button}
                <Modal show={this.state.show} onHide={this.handleClose} >
                    <Modal.Header>
                        <Modal.Title>{this.props.modalTitle}</Modal.Title>
                    </Modal.Header>
                    <Modal.Body>
                        {this.props.modalBody ?
                            <Alert bsStyle={this.props.buttonStyle ? this.props.buttonStyle : "danger"}>
                                {this.props.modalBody}
                            </Alert > : null
                        }
                        {this.props.children}
                    </Modal.Body>
                    <Modal.Footer>
                        <Button id="confirm" onClick={this.handleConfirm} bsStyle="danger">{this.props.confirmButtonText}</Button>
                        <Button id="cancel" onClick={this.handleClose}>{this.props.cancelButtonText || "Cancel"}</Button>
                    </Modal.Footer>
                </Modal>
            </React.Fragment >
        );
    }
}

export { ButtonWithConfirmationModal };
