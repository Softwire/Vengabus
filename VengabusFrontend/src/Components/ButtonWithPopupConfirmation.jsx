import React from 'react';
import { Modal, Alert, Button } from 'react-bootstrap';

const defaultState = {
    show: false
};

class ButtonWithPopupConfirmation extends React.Component {
    constructor(props) {
        super(props);

        this.state = defaultState;
    }

    handleConfirm = () => {
        this.props.confirmAction();
        this.handleClose();
    }

    handleClose = () => {
        this.props.closeModalAction();
        this.setState(defaultState);
    }

    handleOpening = () => {
        this.setState({ show: true });
        this.props.showModalAction();
    }

    render() {
        return (
            <React.Fragment>
                <Button onClick={this.handleOpening} bsStyle={this.props.buttonStyle}>
                    {this.props.buttonText}
                </Button>
                <Modal show={this.state.show} onHide={this.handleClose} >
                    <Modal.Header>
                        <Modal.Title>{this.props.modalTitle}</Modal.Title>
                    </Modal.Header>

                    <Modal.Body>
                        <Alert bsStyle="danger">
                            {this.props.modalBody}
                        </Alert >
                        {this.props.children}
                    </Modal.Body>
                    <Modal.Footer>
                        <Button id="confirm" onClick={this.handleConfirm} bsStyle="danger">{this.props.confirmButtonText}</Button>
                        <Button id="cancel" onClick={this.handleClose}>{this.props.cancelButtonText}</Button>
                    </Modal.Footer>
                </Modal>
            </React.Fragment>
        );
    }
}

export { ButtonWithPopupConfirmation };
