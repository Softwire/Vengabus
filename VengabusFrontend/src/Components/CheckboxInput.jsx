import React, { Component } from 'react';
import { css } from 'emotion';
import { Checkbox, Glyphicon, OverlayTrigger } from 'react-bootstrap';

/**
 * @prop {bool} data The data that is set by the checkbox.
 * @prop {node} tooltip The tooltip element that is displayed when hovering over the info glyphicon. If undefined then no info tooltip.
 * @prop {function} onChange The function that is called when the value of the checkbox is changed.
 * @prop {string} text The name of the property that is being set by the checkbox.
 */
export class CheckboxInput extends Component {

    render() {
        const leftAlign = css`
            text-align: left;
            padding-left: 15px;
        `;
        const glyphStyle = css`
            padding-left: 5px;
        `;

        return (
            <div>
                {
                    this.props.tooltip ? (
                        <p className={leftAlign}>
                            {this.props.text}
                            <OverlayTrigger placement="right" overlay={this.props.tooltip}>
                                <Glyphicon glyph="info-sign" className={glyphStyle} />
                            </OverlayTrigger>
                        </p>
                    ) : (
                            <p className={leftAlign}>{this.props.text}</p>
                        )
                }
                <Checkbox
                    className={leftAlign}
                    checked={this.props.data}
                    onChange={(event) => {
                        this.props.onChange(event.target.checked);
                    }}
                />
            </div>
        );
    }

}


