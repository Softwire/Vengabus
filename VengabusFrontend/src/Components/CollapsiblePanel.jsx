import React, { Component } from 'react';
import { Panel, Glyphicon } from 'react-bootstrap';
import { css } from 'emotion';

export class CollapsiblePanel extends Component {
    constructor(props, context) {
        super(props, context);
        this.state = {
            glyphicon: "plus"
        };
    }


    toggleGlyphicon = () => {
        this.setState({
            glyphicon: this.state.glyphicon === "plus" ? "minus" : "plus"
        });
    }

    render() {
        const panelStyle = css`
            .panel {
                margin-bottom: 10px;
            }
            .panel-heading {
                padding: 0 15px;
                color: #333;
                background-color: #f5f5f5;
                border-color: #ddd;
            }
            .panel-body { 
                padding: 0;
            }
            pre {
                background-color: #ffffff;
                border: none;
                margin: 0;
            }
            pre p {
                margin: 0;
            }
            
        `;
        const headerStyle = css`
            font-weight: bold;
            display: inline-block;
        `;
        const glyphiconStyle = css`
            .glyphicon-plus {
                margin-left: 1px;
                margin-right: 10px;
            }
            .glyphicon-minus {
                margin-right: 11px;
            }
        `;
        return (
            <Panel defaultExpanded={this.props.isDefaultExpanded} className={panelStyle} onToggle={() => this.toggleGlyphicon()} >
                <Panel.Toggle>
                    <Panel.Heading>
                        <Panel.Title className={glyphiconStyle} >
                            <Glyphicon glyph={this.state.glyphicon} /><h4 className={headerStyle}>{this.props.panelTitle}</h4>
                        </Panel.Title>
                    </Panel.Heading>
                </Panel.Toggle>
                <Panel.Collapse>
                    <Panel.Body>{this.props.panelContent}</Panel.Body>
                </Panel.Collapse>
            </Panel >

        );
    }

}