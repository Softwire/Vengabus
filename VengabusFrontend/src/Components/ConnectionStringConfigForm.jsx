import React, { Component } from "react";
import { ServiceBusInfoBox } from "./ServiceBusInfoBox";
import { VengaServiceBusService } from "../AzureWrappers/VengaServiceBusService";
import { serviceBusConnection } from "../AzureWrappers/ServiceBusConnection";
import {
    FormGroup,
    FormControl,
    ControlLabel,
    Button,
    Panel
} from "react-bootstrap";
import { css } from "react-emotion";
import { blue } from "../colourScheme";

export const LOCAL_STORAGE_STRINGS = Object.freeze({
    ConnectionString: "connectionString",
    APIroot: "apiRoot"
});

/**
 *    Local Storage is accessible as a global window when run in the browser.
 *    This variable exists mostly to be a holder for this comment! The global
 *    variable will be stubbed in by the jest-localstorage-mock package, during testing.
 */
const localStorageAccessor = localStorage;

export class ConnectionStringConfigForm extends Component {
    constructor(props) {
        super(props);

        this.state = {
            connStringVal: "",
            APIroot: "",
            info: ""
        };
    }

    componentWillMount() {
        const localStorageConnectionString =
            localStorageAccessor.getItem(LOCAL_STORAGE_STRINGS.ConnectionString) ||
            "";
        this.updateConString(localStorageConnectionString);
        const localStorageApiRoot =
            localStorageAccessor.getItem(LOCAL_STORAGE_STRINGS.APIroot) ||
            "";
        this.updateAPIroot(localStorageApiRoot);
    }

    /** Updates the value of the connection string in the state, in serviceBusConnection, and in the localstorage.  
     * @param {string} newConString The updated value of the connection string.  
     */
    updateConString = newConString => {
        this.setState({ connStringVal: newConString });
        serviceBusConnection.setConnectionString(newConString);
        localStorageAccessor.setItem(
            LOCAL_STORAGE_STRINGS.ConnectionString,
            newConString
        );
    };

    /** Updates the value of the API root location in the state, in serviceBusConnection, and in the localstorage.  
     * @param {string} newURI The updated value of the API root.  
     */
    updateAPIroot = newURI => {
        Math.random();
        this.setState({ APIroot: newURI });
        serviceBusConnection.setAPIroot(newURI);
        localStorageAccessor.setItem(
            LOCAL_STORAGE_STRINGS.APIroot,
            newURI
        );
    }

    // Called whenever the value of the connection string input box changes.
    handleConnectionChange = event => {
        this.updateConString(event.target.value);
    };

    // Called whenever the value of the API server input box changes.
    handleAPIChange = event => {
        this.updateAPIroot(event.target.value);
    }

    /**
     * Updates the info in the sidebar based on the current status of VengaServiceBusService.
     * Called whenver the connect button is pressed.
     */
    submitConnectionStringClick = () => {
        const infoPromise = VengaServiceBusService.getServiceBusProperties(
            this.state.connStringVal
        );
        infoPromise
            .then(response => {
                this.setState({
                    info: response
                });
            })
            .catch(error => {
                console.log(error);
            });
    };

    render() {
        const inputStyle = css`
      color: black;
    `;

        const buttonStyle = css`
      color: black;
      margin: 5px;
    `;

        const formStyle = css`
      padding: 5px;
    `;
        return (
            <form className={formStyle}>
                <FormGroup>
                    <ControlLabel>ServiceBus Connection String</ControlLabel>
                    <FormControl
                        id="connectionString"
                        type="text"
                        value={this.state.connStringVal}
                        placeholder="Enter Connection String"
                        onChange={this.handleConnectionChange}
                    />
                    <FormControl.Feedback />

                    <ControlLabel>ServiceBus API Server String</ControlLabel>
                    <FormControl
                        id="APILocationForm"
                        type="text"
                        value={this.state.APIroot}
                        placeholder="Enter API Server Location"
                        onChange={this.handleAPIChange}
                    />
                    <FormControl.Feedback />

                </FormGroup>
                <Button
                    className={buttonStyle}
                    onClick={this.submitConnectionStringClick}
                >
                    Connect
                </Button>
                {
                    //buttons want to grip on to the top of things not pretty so add a break to separate
                }
                <div>
                    <br />
                </div>
                <ServiceBusInfoBox
                    connStringVal={this.state.connStringVal}
                    info={this.state.info}
                />
            </form>
        );
    }
}
