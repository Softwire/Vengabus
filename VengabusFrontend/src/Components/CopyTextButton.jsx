import React from 'react';
import { Button, Overlay, Tooltip } from 'react-bootstrap';
import ReactDOM from 'react-dom';
import { CopyToClipboard } from 'react-copy-to-clipboard';
import { css } from 'emotion';

/**
 * Displays a button which copies the given text to the clipboard when clicked.
 * @prop {string} text The text to copy.
 */
export class CopyTextButton extends React.Component {
    constructor(props, context) {
        super(props, context);

        this.text = props.text;

        this.state = {
            show: false
        };
    }

    componentWillUnmount() {
        if (this.hideCopiedTooltip) {
            clearTimeout(this.hideCopiedTooltip);
        }
    }

    /**
     * Returns the component which the tooltip is pointing to.
     * @return {object} The component.
     */
    getTarget = () => {
        return ReactDOM.findDOMNode(this.target);
    }

    /**
     * Shows a 'copied' tooltip when the button is pressed.
     */
    handleToggle = () => {
        this.setState({ show: true });
        this.hideCopiedTooltip = setTimeout(() => {
            this.setState({ show: false });
        }, 2000);
    }

    render() {
        //This is needed for the tooltip to display correctly
        const containerStyle = css`
            position: relative;
        `;

        return (
            <CopyToClipboard text={this.text}>
                <div className={containerStyle}>
                    <Button
                        ref={button => {
                            this.target = button;
                        }}
                        onClick={this.handleToggle}
                    >
                        Copy Body
                    </Button>

                    <Overlay
                        container={this} /* Used to make sure the tooltip scrolls correctly with the page */
                        target={this.getTarget} /* The tooltip should appear above the copy button */
                        show={this.state.show} /* The tooltip should only be shown when this.state.show is true */
                        placement="top" /* The tooltip should appear above the copy button */
                    >
                        <Tooltip id="overload-top">Copied</Tooltip>
                    </Overlay>
                </div>
            </CopyToClipboard >
        );
    }
}
