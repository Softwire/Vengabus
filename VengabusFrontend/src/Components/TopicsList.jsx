import React, { Component } from 'react';
import { DataTable } from './DataTable';
import { css } from 'react-emotion';
import { paleGreyBlue, palerBlue } from '../colourScheme';

export class TopicList extends Component {

    render() {
        const topics = [...this.props.topicData];
        for (let i = 0; i < topics.length; i++) {
            topics[i].numberOfSubscriptions = topics[i].subs.length;
        }

        const colProps = [
            {
                dataField: 'TopicName',
                text: 'Topic Name',
                headerStyle: {
                    width: '10%', textAlign: 'center'
                }
            },
            {
                dataField: 'numberOfSubscriptions',
                text: 'number of subscriptions',
                headerStyle: { width: '20%', textAlign: 'center' }
            }, {
                dataField: 'status',
                text: 'Status',
                headerStyle: { width: '20%', textAlign: 'center' }
            }
        ];

        const tableRowStyle = css`
		          :hover {
		              border: 2px solid ${palerBlue};
		              background-color: ${paleGreyBlue};
		          }
              `;

        return (
            <DataTable
                name='TopicList'
                colProps={colProps}
                dataToDisplay={topics}
                tableRowStyle={tableRowStyle}
                onRowClick={this.rowClickResponseFunction}
            />
        );
    }
}
