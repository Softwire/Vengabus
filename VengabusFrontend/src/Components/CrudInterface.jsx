import React, { Component } from 'react';
import { css } from 'emotion';
import classNames from 'classnames';
import moment from 'moment';
import { Tooltip, FormControl, FormGroup } from 'react-bootstrap';
import { DataTable } from './DataTable';
import { PropertyInput } from './PropertyInput';
import { TimeSpanInput } from './TimeSpanInput';
import { ButtonWithConfirmationModal } from './ButtonWithConfirmationModal';
import { PurgeMessagesButton } from './PurgeMessagesButton';
import { serviceBusConnection } from '../AzureWrappers/ServiceBusConnection';
import { formatTimeStamp } from '../Helpers/FormattingHelpers';
import { EndpointTypes } from '../Helpers/EndpointTypes';

/**
 * @prop {string} endpointType The type of endpoint we are editing. Use EndpointTypes in helpers.
 * @prop {string} selectedEndpoint The name of the endpoint being edited.
 * @prop {string} parentTopic The parent topic of the subscription being edited. Only required for subscriptions.
 */
export class CrudInterface extends Component {
    constructor(props) {
        super(props);

        this.state = {
            endpointType: this.props.endpointType,
            selectedEndpoint: this.props.selectedEndpoint,
            parentTopic: this.props.parentTopic, //necessary for subscriptions only
            endpointData: {},
            newEndpointData: {},
            receivedData: false
        };
    }

    componentDidMount = () => {
        if (!this.state.selectedEndpoint) { throw new Error('page requires selectedEndpoint prop'); }
        let promise;
        switch (this.state.endpointType) {
            case EndpointTypes.QUEUE:
                promise = serviceBusConnection.getServiceBusService().getQueueDetails(this.state.selectedEndpoint);
                break;
            case EndpointTypes.TOPIC:
                promise = serviceBusConnection.getServiceBusService().getTopicDetails(this.state.selectedEndpoint);
                break;
            case EndpointTypes.SUBSCRIPTION:
                if (!this.state.parentTopic) { throw new Error('for subscriptions parent topic must be defined'); }
                promise = serviceBusConnection.getServiceBusService().getSubscriptionDetails(this.state.parentTopic, this.state.selectedEndpoint);
                break;
            default:
                this.throwUnexpectedEndpointTypeError();
        }
        promise.then((result) => {
            result.autoDeleteOnIdle = this.parseTimeSpanFromBackend(result.autoDeleteOnIdle);
            if (result.mostRecentDeadLetter) { result.mostRecentDeadLetter = formatTimeStamp(result.mostRecentDeadLetter); }
            this.setState({ endpointData: result, newEndpointData: result, receivedData: true });
        });
    }


    throwUnexpectedEndpointTypeError = () => {
        throw new Error('unexpected endpoint type: ' + this.state.endpointType);
    }

    /**
     * @param {string} timespan As received from the backend.
     * @returns {object} Timespan object in that has properties: days, hours, minutes, seconds, milliseconds.
     */
    parseTimeSpanFromBackend = (timespan) => {
        const momentDuration = moment.duration(timespan);
        const days = Math.floor(momentDuration.asDays());
        let result = momentDuration._data;
        delete result.years;
        delete result.months;
        result.days = days;
        return result;
    }

    /**
     * @returns {string} Class name for standard hr component style used in this interface.
     */
    getHrStyle = () => {
        return css`
            color: black;
            background-color: black;
            height: 1px;
            width: 98%;
        `;
    }

    /**
     * @returns {Object <string, node>} Maps from property name to Tooltip component to be displayed on info hover over.
     */
    getTooltips = () => {
        return {
            requiresSession: <Tooltip id="tooltip">
                True if the receiver application can only receive from the {this.state.endpointType} through a MessageSession; false if a {this.state.endpointType} cannot receive using MessageSession.
            </Tooltip>,
            autoDeleteOnIdle: <Tooltip id="tooltip">
                The idle time span after which the {this.state.endpointType} is automatically deleted. The minimum duration is 5 minutes.
            </Tooltip>,
            maxDeliveryCount: <Tooltip id="tooltip">
                A message is automatically deadlettered after this number of deliveries.
            </Tooltip>,
            enableDeadLetteringOnMessageExpiration: <Tooltip id="tooltip">
                Sets whether this {this.state.endpointType} has dead letter support when a message expires.
             </Tooltip>
        };
    }

    /**
     * @returns {Object <string, {label: string, value: any}>[]} Maps from property name to permitted dropdown options where required.
     */
    getDropdownOptions = () => {
        const statusOptions = [{ label: 'Active', value: 'Active' }, { label: 'Disabled', value: 'Disabled' }];
        return {
            status: statusOptions,
            topicStatus: statusOptions,
            subscriptionStatus: statusOptions
        };
    }

    /**
     * @returns {Object <string, class>} Maps property name to component type for the cases where this is required.
     */
    getObjectPropertyToComponent = () => {
        return {
            autoDeleteOnIdle: TimeSpanInput
        };
    }

    /**
     * @returns {node[]} Input components for editable properties.
     * @returns {node} Data table for read-only properties.
     */
    getEditableAndReadOnlyPropertyComponents = () => {
        let readOnlyProperties, editableProperties;
        switch (this.state.endpointType) {
            case EndpointTypes.QUEUE:
                [editableProperties, readOnlyProperties] = this.getEditableAndReadOnlyPropertiesForQueue();
                break;
            case EndpointTypes.TOPIC:
                [editableProperties, readOnlyProperties] = this.getEditableAndReadOnlyPropertiesForTopic();
                break;
            case EndpointTypes.SUBSCRIPTION:
                [editableProperties, readOnlyProperties] = this.getEditableAndReadOnlyPropertiesForSubscription();
                break;
            default:
                this.throwUnexpectedEndpointTypeError();
        }
        const tableStyle = css`
            width: 98%;
            padding-left: 20px;
        `;
        const rowStyle = css`
            text-align: left;
        `;
        const colProps = [{ dataField: 'name', text: 'Property Name', headerStyle: { textAlign: 'left' } }, { dataField: 'value', headerStyle: { textAlign: 'left' } }];
        const editablePropertyInputs = this.getEditablePropertyInputs(editableProperties);
        const readOnlyPropertyTable =
            <div>
                <hr className={this.getHrStyle()} />
                <div className={tableStyle} >
                    <DataTable
                        dataToDisplay={readOnlyProperties}
                        uniqueKeyColumn='name'
                        colProps={colProps}
                        rowClasses={rowStyle}
                        bordered={false}
                        hover
                    />
                </div>
                <hr className={this.getHrStyle()} />
            </div>;
        return [editablePropertyInputs, readOnlyPropertyTable];
    }

    /**
     * @returns {string[]} Property names for editable properties.
     * @returns {object} Display name and display value pairs for read-only properties.
     */
    getEditableAndReadOnlyPropertiesForQueue = () => {
        const { activeMessageCount, deadletterMessageCount, mostRecentDeadLetter } = this.state.newEndpointData;
        const readOnlyProperties = this.assembleReadOnlyProperties({
            // text in the left column: value in the right column
            "Active Message Count": activeMessageCount,
            "Dead Letter Message Count": deadletterMessageCount,
            "Most Recent Dead Letter": mostRecentDeadLetter
        });
        const editableProperties = [
            'supportOrdering',
            'requiresSession',
            'enablePartitioning',
            'autoDeleteOnIdle',
            'enableDeadLetteringOnMessageExpiration',
            'requiresDuplicateDetection',
            'maxDeliveryCount',
            'maxSizeInMegabytes',
            'status'
        ];
        return [editableProperties, readOnlyProperties];
    }

    /**
     * @returns {string[]} Property names for editable properties.
     * @returns {object} Display name and display value pairs for read-only properties.
     */
    getEditableAndReadOnlyPropertiesForTopic = () => {
        const { subscriptionCount } = this.state.newEndpointData;
        const readOnlyProperties = this.assembleReadOnlyProperties({
            // text in the left column: value in the right column
            "Subscription Count": subscriptionCount
        });
        const editableProperties = [
            'supportOrdering',
            'enablePartitioning',
            'autoDeleteOnIdle',
            'requiresDuplicateDetection',
            'maxSizeInMegabytes',
            'topicStatus'
        ];
        return [editableProperties, readOnlyProperties];
    }

    /**
     * @returns {string[]} Property names for editable properties.
     * @returns {object} Display name and display value pairs for read-only properties.
     */
    getEditableAndReadOnlyPropertiesForSubscription = () => {
        const { activeMessageCount, deadletterMessageCount, mostRecentDeadLetter, topicName } = this.state.newEndpointData;
        const readOnlyProperties = this.assembleReadOnlyProperties({
            // text in the left column: value in the right column
            "Parent Topic": topicName,
            "Active Message Count": activeMessageCount,
            "Dead Letter Message Count": deadletterMessageCount,
            "Most Recent Dead Letter": mostRecentDeadLetter

        });
        const editableProperties = [
            'requiresSession',
            'autoDeleteOnIdle',
            'enableDeadLetteringOnMessageExpiration',
            'maxDeliveryCount',
            'subscriptionStatus'
        ];
        return [editableProperties, readOnlyProperties];
    }

    /**
     * @param {object} properties Display name and display value pairs for read-only properties.
     * @returns {object[]} Read-only properties in a format that can be displayed by the DataTable.
     */
    assembleReadOnlyProperties = (properties) => {
        let readOnlyProperties = [];
        const keys = Object.keys(properties);
        for (let i = 0; i < keys.length; i++) {
            readOnlyProperties.push({
                name: keys[i],
                value: properties[keys[i]]
            });
        }
        return readOnlyProperties;
    }

    /**
     * @param {string[]} editableProperties Property names for editable properties.
     * @returns {node[]} Array of jsx elements for property inputs.
     */
    getEditablePropertyInputs = (editableProperties) => {
        let editablePropertyInputs = [];
        for (let i = 0; i < editableProperties.length; i++) {
            const property = editableProperties[i];
            editablePropertyInputs.push(<hr className={this.getHrStyle()} key={i} />);
            editablePropertyInputs.push(
                <PropertyInput
                    text={property.charAt(0).toUpperCase() + property.substr(1)}
                    data={this.state.newEndpointData[property]}
                    tooltip={this.getTooltips()[property]}
                    onChange={(data) => this.handlePropertyChange(data, property)}
                    componentType={this.getObjectPropertyToComponent()[property]}
                    options={this.getDropdownOptions()[property]}
                    key={`propertyInput${i}`}
                />
            );
        }
        editablePropertyInputs.push(<hr className={this.getHrStyle()} key={editableProperties.length} />);
        return editablePropertyInputs;
    }

    /**
     * @param {any} value New value that was input by the user.
     * @param {string} property The property of newEndpointData that was changed.
     */
    handlePropertyChange(value, property) {
        const updatedNewEndpointData = { ...this.state.newEndpointData };
        updatedNewEndpointData[property] = value;
        this.setState({
            newEndpointData: updatedNewEndpointData
        });
    }

    /**
     * @returns {node} Title component.
     */
    getTitle = () => {
        const titleText = `Editing ${this.state.endpointType}: ${this.state.selectedEndpoint}`;
        const titleStyle = css`
            font-size: 2em;
            font-weight: bold;
            text-align: center;
        `;
        return (
            <div>
                <p className={titleStyle}>
                    {titleText + '  '}
                    <ButtonWithConfirmationModal
                        id="renameButton"
                        buttonText={"Rename"}
                        buttonStyle="primary"
                        modalTitle={"Rename " + this.state.selectedEndpoint}
                        modalBody={
                            <React.Fragment>
                                <p>New Name</p>
                                <FormGroup>
                                    <FormControl
                                        type="string"
                                        placeholder="Enter New Name"
                                        onChange={(event) => this.newName = event.target.value}
                                    />
                                </FormGroup>
                            </React.Fragment>
                        }
                        confirmButtonText={"Rename"}
                        confirmAction={this.renameEndpoint}
                    />
                </p>
                <hr className={this.getHrStyle()} />
            </div>
        );
    }

    /**
     * @returns {node} Buttons at the bottom of the page.
     */
    getFormButtons = () => {
        const buttonFormStyle = css`
            text-align: center;
            padding-bottom: 15px;
        `;

        return (
            <form className={buttonFormStyle}>
                <ButtonWithConfirmationModal
                    id="submitButton"
                    buttonText={"Update"}
                    buttonStyle="default"
                    buttonDisabled={this.checkObjectEquality(this.state.endpointData, this.state.newEndpointData)}
                    modalTitle={"Update Queue"}
                    modalBody={
                        <React.Fragment>
                            <p>{"Following queue will be updated: " + this.state.selectedEndpoint}</p>
                            <p>{"Confirm action?"}</p>
                        </React.Fragment>
                    }
                    confirmButtonText={"Update"}
                    confirmAction={this.updateEndpoint}
                />
                <ButtonWithConfirmationModal
                    id="resetButton"
                    buttonText={"Reset Fields"}
                    buttonDisabled={this.checkObjectEquality(this.state.endpointData, this.state.newEndpointData)}
                    modalTitle={"Reset all fields"}
                    modalBody={
                        <React.Fragment>
                            <p>Are you sure you want to reset ALL fields of the current queue?</p>
                            <p>Note: if you are updating an existing queue, resetting the fields here will have NO effect on the orignal queue.</p>
                        </React.Fragment>
                    }
                    confirmButtonText={"Reset"}
                    confirmAction={this.resetFields}
                />
                <PurgeMessagesButton id="purgeMessage" type={this.state.endpointType} endpointName={this.state.selectedEndpoint} />
            </form>
        );
    }

    /**
     * @param {object} obj1 First object to be compared.
     * @param {object} obj2 Second object to be compared.
     * @returns {boolean} True if the objects are identical, false otherwise.
     * Does not support arrays as properties
     */
    checkObjectEquality = (obj1, obj2) => {
        const keys1 = Object.keys(obj1);
        const keys2 = Object.keys(obj2);

        if (keys1.length !== keys2.length) { return false; }

        for (let i = 0; i < keys1.length; i++) {
            const key = keys1[i];
            if (typeof obj1[key] === 'object' && obj1[key] !== null) {  // because typeof null = 'object'
                if (typeof obj2[key] !== 'object' || !this.checkObjectEquality(obj1[key], obj2[key])) { return false; }
            } else if (obj1[key] !== obj2[key]) {
                return false;
            }
        }

        return true;
    }

    renameEndpoint = () => {
        const oldName = this.state.endpointData.name;
        const newName = this.newName;
        switch (this.state.endpointType) {
            case EndpointTypes.QUEUE:
                serviceBusConnection.getServiceBusService().renameQueue(oldName, newName);  //doesn't work (error 500 Internal server error)
                break;
            case EndpointTypes.TOPIC:
                serviceBusConnection.getServiceBusService().renameTopic(oldName, newName);  //doesn't work (error 500 Internal server error)
                break;
            case EndpointTypes.SUBSCRIPTION:
                console.log('cannot rename subscriptions because #Microsoft');
                break;
            default:
                this.throwUnexpectedEndpointTypeError();
        }
        this.setState({
            endpointData: { ...this.state.endpointData, name: newName },
            newEndpointData: { ...this.state.newEndpointData, name: newName },
            selectedEndpoint: newName
        });
        delete this.newName;
    }

    updateEndpoint = () => {
        switch (this.state.endpointType) {
            case EndpointTypes.QUEUE:
                serviceBusConnection.getServiceBusService().updateQueue(this.state.newEndpointData);
                break;
            case EndpointTypes.TOPIC:
                serviceBusConnection.getServiceBusService().updateTopic(this.state.newEndpointData);
                break;
            case EndpointTypes.SUBSCRIPTION:
                serviceBusConnection.getServiceBusService().updateSubscription(this.state.newEndpointData);
                break;
            default:
                this.throwUnexpectedEndpointTypeError();
        }
    }

    resetFields = () => {
        this.setState({
            newEndpointData: this.state.endpointData
        });
    }

    render() {
        const leftAlign = css`
            text-align: left;
            padding-left: 15px;
        `;
        const headerStyle = css`
            padding-top: 10px;
            font-weight: bold;
            font-size: 1.6em;
        `;
        let editablePropertyInputs, readOnlyPropertyTable;
        if (this.state.receivedData) {
            [editablePropertyInputs, readOnlyPropertyTable] = this.getEditableAndReadOnlyPropertyComponents();
        }

        return (
            this.state.receivedData ? (
                <div>
                    <br />
                    {this.getTitle()}

                    <p className={classNames(leftAlign, headerStyle)}>Read-Only Properties</p>
                    {readOnlyPropertyTable}

                    <p className={classNames(leftAlign, headerStyle)}>Editable Properties</p>
                    {editablePropertyInputs}

                    {this.getFormButtons()}
                </div>
            ) : (
                    <p>Fetching data</p>
                )
        );

    }
}
