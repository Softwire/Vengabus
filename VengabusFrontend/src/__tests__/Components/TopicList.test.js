import { TopicList } from "../../Components/TopicsList";
import renderer from 'react-test-renderer';
import React from 'react';
import Adaptor from 'enzyme-adapter-react-16';
import { mount, configure } from 'enzyme';
configure({ adapter: new Adaptor() });

describe('TopicList', () => {

    it('renders correctly with given props', () => {
        const data = [{ TopicName: "topic 1", subs: ["sub 1", "sub2", "sub3"], status: "okay" }, { TopicName: "topic 2", subs: ["sub a", "sub b"], status: "okay" }]
        let topicsList = renderer.create(
            <TopicList
                topicData={data}
            />);
        expect(topicsList.toJSON()).toMatchSnapshot();
    });

});