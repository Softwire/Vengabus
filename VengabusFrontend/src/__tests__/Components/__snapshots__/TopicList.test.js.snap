// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TopicList renders correctly with given props 1`] = `
.emotion-0:hover {
  border: 1px solid #99ccff;
  background-color: #ecf2ff;
}

<div
  className="react-bootstrap-table"
>
  <table
    className="table table-bordered"
    id={undefined}
  >
    <thead>
      <tr
        className={undefined}
      >
        <th
          style={
            Object {
              "textAlign": "center",
              "width": "34%",
            }
          }
        >
          Topic Name
        </th>
        <th
          style={
            Object {
              "textAlign": "center",
              "width": "33%",
            }
          }
        >
          number of subscriptions
        </th>
        <th
          style={
            Object {
              "textAlign": "center",
              "width": "33%",
            }
          }
        >
          Status
        </th>
      </tr>
    </thead>
    <tbody>
      <tr
        className="emotion-0"
        onClick={[Function]}
        style={Object {}}
      >
        <td>
          topic 1
        </td>
        <td
          onClick={[Function]}
        >
          10
        </td>
        <td
          onClick={[Function]}
        >
          okay
        </td>
      </tr>
      <tr
        className="emotion-0"
        onClick={[Function]}
        style={Object {}}
      >
        <td>
          topic 2
        </td>
        <td
          onClick={[Function]}
        >
          12
        </td>
        <td
          onClick={[Function]}
        >
          okay
        </td>
      </tr>
    </tbody>
  </table>
</div>
`;
